Feature/unit test missing assert

gardian12 opened the pull request:
Hello,
I've added a new check, which raises a finding if no assertion at a testing method was found.
At least a static method call of `CL_ABAP_UNIT_ASSERT` should be added to each testing method to achieve a valuable unit test.

gardian12 committed:
added new check "unit test missing assert"

gardian12 committed:
added new check "unit test missing assert" - impl

gardian12 committed:
Update src/checks/y_check_missing_ut_assert.clas.abap

removed comments

Co-authored-by: abaplint[bot] <24845621+abaplint[bot]@users.noreply.github.com>

bjoern-jueliger-sap commented:
Don't worry, contributions to this repository are still welcome (although of course we don't want to add new checks here and then never add it to the Cloud version)!

Before we get into the code itself, just to clarify:

This check is supposed to enforce the rule about [not adding tests without asserts to improve the coverage](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#dont-obsess-about-coverage), right?

If so, I think we should perhaps be a bit more clear about its goal, since the current assertion in the [documentation](https://github.com/gardian12/code-pal-for-abap/blob/cbb52dd55b4e5df2f018e50f11ec670b771973f7/docs/checks/unit_test_missing_assert.md?plain=1#L7) - "A unit test that lacks any assertions or validation of expected behavior is fundamentally meaningless." - is not actually true: You can write a unit test that is supposed to test that a certain unit doesn't crash:

```abap
method test_no_exception.
  data(cut) = new class_supposed_to_be_robust( ).
  cut->do( garbage_input ).
endmethod.
```

And this would test that `cut->do` does not crash when given `garbage_input` (by throwing either runtime errors or exceptions). Of course one can argue that one should actually test the full behaviour of the code (i.e. the return value or whatever side effects it is supposed to have in this case), but it's not meaningless (and some people do this also e.g. because they don't care about what happens with garbage input, just that it doesn't crash).

I don't think a static check can ever really detect these kinds of tests, but it shouldn't claim that they're useless - just that it's better to add an explicit assert (this also falls into making [tests readable](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#readability-rules)). However, users will probably want a pseudo comment for these cases, anyway.

bjoern-jueliger-sap commented:
I'll have a look at it. The parser should not delete pseudo comments in test code any more than it does in real code, but the problem might be the placement of the comment in relation to the finding

bjoern-jueliger-sap committed:
Add pseudo comment

The CodePal always looks for the pseudo comment at the position of the finding. Since we position the finding on the first line of the method implementation, the unit test also needs to position it at the first line, then it works.

bjoern-jueliger-sap commented:
I've also run into a different case where the check reports a false positive that's a bit more problematic than the artificial case above since if one follows the [write custom asserts](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#write-custom-asserts-to-shorten-code-and-avoid-duplication) rule, _every_ test method one writes will get a finding from this check.

The check currently only looks for the assert in the test method itself, but not in any called methods, so e.g. this code

```abap
class actual_tests definition for testing
    risk level harmless
    duration short
    inheriting from tests.
  PRIVATE SECTION.

    METHODS custom_assert FOR TESTING RAISING cx_static_check.
    METHODS assert_tbl_eq_ignoring_order
      IMPORTING
        value(exp_tab) type standard table
        value(act_tab) type standard table.
endclass.

CLASS actual_tests IMPLEMENTATION.

  METHOD custom_assert.
    types ty_tab type standard table of i with empty key.
    data(my_expected_table) = value ty_tab( ( 1 ) ( 2 ) ).
    data(my_actual_table) = value ty_tab( ( 2 ) ( 1 ) ( 3 ) ).

    assert_tbl_eq_ignoring_order(
      exp_tab = my_expected_table
      act_tab = my_actual_table ).
  ENDMETHOD.

  METHOD assert_tbl_eq_ignoring_order.
    loop at exp_tab assigning field-symbol(<exp_tab_line>).
      data(exp_idx) = sy-tabix.
      data(act_idx) = line_index( act_tab[ table_line = <exp_tab_line> ] ).
      if act_idx > 0.
        delete act_tab index act_idx.
        delete exp_tab.
        continue.
      else.
        cl_abap_unit_assert=>fail(
          msg = |Line index { exp_idx } of expected table not in actual table.| ).
      endif.
    endloop.
    if lines( act_tab ) > 0.
      cl_abap_unit_assert=>fail(
        msg = |{ lines( act_tab ) } lines of actual table not in expected table.| ).
    endif.
  ENDMETHOD.

ENDCLASS.
```

gets a finding in `custom_assert` and would get another finding in every other method that reuses the custom assert method that compares tables without order.

In the more modern cloud infrastructure it would be comparatively simple to analyze also the called methods for the assert (though one has to be careful with avoiding recursion etc. when doing that), but in this older SCAN-based infrastructure that's a bit of a pain.

gardian12 commented:
Hello,
I've got your point.
Considering the implementation of custom assertions has to follow this approach [IF_CONSTRAINT](https://help.sap.com/doc/saphelp_snc700_ehp01/7.0.1/de-DE/73/a417fb3e80462980dd4caf0a653041/content.htm).
To be honest, I haven't used it so far, but I think expecting this could be carried out be check. So false-positives shouldn't be a problem. Will extend the implementation this way and have a look if it works.
Many thanks so far for solving the pragma issue.

gardian12 commented:
Hello @bjoern-jueliger-sap, it's been a while. I wanted to follow up on how we want to proceed with this pull request?
The only open point is the pragma option, which didn't work, so I had to disable it.
So how we do proceed?

bjoern-jueliger-sap commented:
Sorry for the late reply - I'm still a bit unsure about the IF_CONSTRAINT thing. The style guide doesn't currently recommend using that technique, and just writes an assert without it in its example. Generally we try to avoid discrepancies between what the CodePal checks report and what the guide recommends. In sample code bases I also don't see a lot of use of IF_CONSTRAINT, even if the [official SAP documentation](https://help.sap.com/docs/ABAP_PLATFORM_NEW/ba879a6e2ea04d9bb94c7ccd7cdac446/73a417fb3e80462980dd4caf0a653041.html?locale=en-US&version=LATEST) still seems to recommend it.
